### PR TITLE
fix: use detect-newline to detect newline instead of default \n

### PR DESCRIPTION
**Pull request by tuananh**
use `detect-newline` to detect newline instead of using the default `\n` which may cause npm to find that `packagejson` string to be different even though there is no change. (causing by the `\n` newline)

### CODE REVIEW

File: lib/install/save.js [hunk 1]
```diff
@@ -122,7 +124,7 @@ function savePackageJson (tree, next) {
       tree.package.bundleDependencies = deepSortObject(bundle)
     }
 
-    var json = JSON.stringify(tree.package, null, indent) + '\n'
+    var json = JSON.stringify(tree.package, null, indent) + newline
```
Comment (zkat) [inline lib/install/save.js]: This does not replace all the newlines in the JSON file. I'm pretty sure you're gonna have to `JSON.stringify(...).replace(/\n/g, newline)` here, not just append the final one.

File: lib/install/save.js [hunk 2]
```diff
@@ -3,6 +3,7 @@
 const createShrinkwrap = require('../shrinkwrap.js').createShrinkwrap
 const deepSortObject = require('../utils/deep-sort-object.js')
 const detectIndent = require('detect-indent')
+const detectNewline = require('detect-newline')
```
Comment (zkat) [inline lib/install/save.js:6]: This is not the only file that needs you to do this. You'll also have to add this to `lib/shrinkwrap.js` and `lib/version.js`, which both write out json files.

**Review by zkat — CHANGES_REQUESTED** [review]
Thanks for taking the time and initiative to do this!! There's a bit more that needs to go into this patch for it to be ready, and I'm honestly fine not having test for a thing like this, provided the regular test suite continues passing.

File: lib/install/save.js (hunk 2 above)
Comment (tuananh) [reply lib/install/save.js:6]: i updated the PR to cover `lib/shrinkwrap.js` and `lib/version.js` as well.

**Review by zkat — APPROVED** [review]
LGTM! Thanks a bunch for this patch!

**Comment by nwoltman** [comment]
#18943 from 4 months ago solved this in IMO a cleaner way and included tests. Guess you didn't see it...

**Pull request merged**